qemu/tests/hugepage_specify_node.py

```diff
@@ -0,0 +1,81 @@
+import logging
+
+from avocado.utils import memory
+
+from virttest import env_process
+from virttest import error_context
+from virttest import utils_misc
+
+from virttest.utils_numeric import normalize_data_size
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Qemu allocate hugepage from specify node.
+    Steps:
+    1) Setup total of 4G mem hugepages for specify node.
+    2) Setup total of 1G mem hugepages for idle node.
+    3) Mount this hugepage to /mnt/kvm_hugepage.
+    4) Boot guest only allocate hugepage from specify node.
+    5) Check the hugepage used from every node.
+    :params test: QEMU test object.
+    :params params: Dictionary with the test parameters.
+    :params env: Dictionary with test environment.
+    """
+    hugepage_size = memory.get_huge_page_size()
+    mem_size = int(normalize_data_size("%sM" % params["mem"], "K"))
+    idle_node_mem = int(normalize_data_size("%sM" % params["idle_node_mem"], "K"))
+
+    error_context.context("Get host numa topological structure.", logging.info)
+    host_numa_node = utils_misc.NumaInfo()
+    node_list = host_numa_node.online_nodes_withcpumem
+    pre_hugepages_free = {}
+    for node_id in node_list:
+        error_context.base_context("Check preprocess HugePages Free on host "
+                                   "numa node %s." % node_id, logging.info)
+        node_memfree = int(host_numa_node.read_from_node_meminfo(node_id,
+                                                                 "MemFree"))
+        value = int(host_numa_node.read_from_node_meminfo(node_id,
+                                                          "HugePages_Free"))
+        pre_hugepages_free[node_id] = value
+        if node_memfree < mem_size:
+            node_list.remove(node_id)
```

Comment:
The node_id only need to be removed for L49.  For L57 and L72,  should keep the node in the list.

Reply:
Some nodes without memory, so L57 cannot keep the node in the list.

Reply:
Do you mean to add a new judgment to update node_list on L57 ?

Reply:
The node removed in this line is not without memory, it gets removed because its free memory is less than mem_size in L42..

Reply:
I understand, but I mean L57 needs memory with mem greater than 1G to allocate large pages of free nodes.
so you mean to add a new judgment to update node_list on L57 ?

Reply:
I would assume a node with memory would have 1G memory at least. But adding a new judgement would be fine.

Reply:
ok, corrected.